### ISSUE TITLE
kubectl scale should use the scale subresource

### ISSUE THREAD

**Issue by nikhiljindal** [kubernetes/kubernetes#29698]
`$kubectl scale --replicas=3 deployments/my-deployment --v=9` gives

```
curl -k -v -XPUT  -H "User-Agent: kubectl/v1.4.0 (linux/amd64) kubernetes/2f545e1" -H "Accept: application/json, */*" -H "Content-Type: application/json" -H "Authorization: Basic YWRtaW46ZkVSeVcxbUl5cTdvRHFVNg==" https://8.35.198.5/apis/extensions/v1beta1/namespaces/default/deployments/frontend
```

It is using PUT. It should just use the deployments/scale endpoint.

cc @pwittrock @janetkuo

**Comment by fraenkel** [comment]
It seems the client 1.4 generated code didn't carry over a completed ScaleExpansion interface. Just want to make sure we are ok with a slight imbalance with client API usage between unversioned and 1.4.

**Comment by nikhiljindal** [comment]
That seems like a bug.
I see that both [unversioned](https://github.com/kubernetes/kubernetes/blob/f2ddd60eb9e7e9e29f7a105a9a8fa020042e8e52/pkg/client/clientset_generated/internalclientset/typed/extensions/unversioned/scale_expansion.go#L26) and [release_1_3](https://github.com/kubernetes/kubernetes/blob/f2ddd60eb9e7e9e29f7a105a9a8fa020042e8e52/pkg/client/clientset_generated/release_1_3/typed/extensions/v1beta1/scale_expansion.go#L26) have Get and Update in ScaleExpansion, while [release_1_4](https://github.com/kubernetes/kubernetes/blob/fdd239203562ce9a8b71e8837ff67a93a487fd29/pkg/client/clientset_generated/release_1_4/typed/extensions/v1beta1/generated_expansion.go#L31) does not. Feel free to send a PR to fix that.
cc @caesarxuchao

**Comment by caesarxuchao** [comment]
> I see that both unversioned and release_1_3 have Get and Update in ScaleExpansion, while release_1_4 does not. Feel free to send a PR to fix that.

That's a bug. The expansion files need to be manually copied. I'll find time to automate the process next month. Feel free to grab it and send a PR.

But kubectl does not use the generated clientset, does it?

**Comment by fraenkel** [comment]
@caesarxuchao No it doesn't. Just wanted to make sure there wasn't some other reason for it being missing. I will do a separate PR to get that over in 1.4

**Comment by soltysh** [comment]
It is a bug and see #22308 for reasons why it happened.

**Comment by soltysh** [comment]
The PR was reverted, I'll re-open until it gets solved.

**Comment by fraenkel** [comment]
I noticed that I wasn't sending back the updated resource version. That is a bug but not sure if it was causing the issue seen since there is no information behind the failure.

**Comment by janetkuo** [comment]
We need `statefulset/scale` endpoint: #46005

**Comment by soltysh** [comment]
Most importantly we need to have a [reasonable dynamic and generic client](https://github.com/kubernetes/kubernetes/pull/32523#issuecomment-267050916) for all scale operation. Currently we have `Scale` resource existing in every API group that supports scaling. There should be just one to rule them all :)

**Comment by bgrant0607** [comment]
I would like local mutations work for setting the replicas fields, as discussed in #21648. If we want kubectl scale to use the scale subresource, which seems reasonable, we could also make a `kubectl set replicas` command.

cc @kubernetes/sig-cli-proposals @pwittrock @alexandercampbell

**Comment by foxish** [comment]
We are blocked on this. Tracking the items that need to be resolved first:

- [ ] Subresource kind determination (https://github.com/kubernetes/kubernetes/pull/49971)
- [ ] Implementation of a generic scale client.

cc @DirectXMan12 @deads2k

**Comment by nikhita** [comment]
For the record, generic scale client has been implemented: https://github.com/kubernetes/kubernetes/pull/53743.

**Comment by DirectXMan12** [comment]
and PRs to make kubectl scale aware of that are being finished up: #58298

**Comment by janetkuo** [comment]
This is fixed.
